Lit for polymer users article

arthurevans opened the pull request:
Please review for **content**.

We're waiting on an actual article format, this includes minimal changes to add an article type with TOC (top only ATM) and the param list format from the docs pages. This is just for review purposes, and it leaves a number of weird visual glitches (like permalink anchors that are on a separate line, some weird formatting on the top TOC). So please ignore CSS glitches as we'll be throwing that stuff out. (Markdown formatting errors are fair game, though.)

aomarks commented on packages/lit-dev-content/site/blog/2022-07-19-lit-for-polymer-users.md:
```diff
@@ -0,0 +1,935 @@
+  - arthur-evans
+---
+
+Lit is a successor to the Polymer library. If you have a project built with Polymer and want to migrate it to Lit, or if you're familiar with Polymer and want to know how it compares to Lit, this is the right document for you.
```
Since we're no longer using the "Polymer project" brand, and have disassociated Lit from it, would it make sense to start referring to Polymer simply as "Polymer" again, instead of "the Polymer library"? It doesn't really feel like the "Polymer library" name caught on in our community.

arthurevans replied:
I'm not sure about dropping it altogether; but for the sake of brevity I use "Polymer" in most of this article. I updated the text to be more consistent, and only use the full name on the first reference.

sorvell commented on packages/lit-dev-content/site/blog/2022-07-19-lit-for-polymer-users.md:
```diff
@@ -0,0 +1,935 @@
+
+*   A facility for the child to automatically fire change events when properties change (`notify: true`).
+
+This last item is the most problematic. Components fire change events for any change to a property, and each change event is handled synchronously.  Widespread use of two-way binding across an entire application can make it hard to reason about data flow and the order in which components update themselves. A change can propagate both up and down the tree, potentially leading to components being u ...
```
Maybe add something like...
Ideally an event is a discrete signal intentionally sent to communicate an explicit change that isn't otherwise easily observable. Sending an event as a side-effect of setting a property as Polymer does makes the communication potentially redundant and implicit. This implicit behavior, in partcular, can make data flow hard to understand.

arthurevans replied:
Added this with a few edits. Cut some text that was less on-point.

sorvell commented on packages/lit-dev-content/site/blog/2022-07-19-lit-for-polymer-users.md:
```diff
@@ -0,0 +1,935 @@
+
+Like Polymer, Lit does dirty checking when properties change to avoid performing unnecessary work. This can lead to issues if you have a property that holds an object or array. If you mutate the object or array, Lit won't detect a change.
+
+In most cases, the best way to avoid these issues is to use immutable data patterns, so that you always assign a new object or array value instead of mutating an existing object or array.
```
The rules for Polymer and Lit are the same if you don't include `this.set` so I think that needs to be mentioned here.

arthurevans replied:
Mention `this.set`? Or mention that the rules are the same?

arthurevans replied:
Mentioned rules are basically the same, added a paragraph mentioning `this.set` and friends (which will probably need to be replaced with immutable patterns when moving to Lit).

sorvell commented on packages/lit-dev-content/site/blog/2022-07-19-lit-for-polymer-users.md:
````diff
@@ -0,0 +1,935 @@
+Polymer:
+
+```js
+  static get template() {
````
Include:

```
_formatName(given, family) {
  return `${given} ${family}`;
}
```

sorvell commented on packages/lit-dev-content/site/blog/2022-07-19-lit-for-polymer-users.md:
````diff
@@ -0,0 +1,935 @@
+```js
+render() {
+  html`
+      My name is <span>${this._formatName(given, family)}</span>
````
Then this is sooo much better.

```suggestion
      My name is <span>${this.given} ${this.family}</span>
```

arthurevans replied:
In this case both Polymer and Lit could simply insert two strings. Calling a function is more interesting because, for example, formatting the name might be locale-specific. And it's a very easy pattern to translate.

arthurevans replied:
I'm not making these two changes because this section is specifically showing how to use an inline function.

sorvell reviewed (changes requested):
This is really comprehensive. Great job! Suggestions are mostly minor.

arthurevans commented on packages/lit-dev-content/site/blog/2022-07-19-lit-for-polymer-users.md:
```diff
@@ -0,0 +1,935 @@
+
+There are many ways to replace two-way bindings. On the more complex side, you can use a state container like Redux. Individual components can subscribe for updates to the state, and dispatch actions to update the state. This is a great way to go, but it may require a lot of work.
+
+If you're just looking to migrate an existing Polymer application to Lit, you may just want to replace the two-way bindings with code that serves a similar function. There are two steps to this:
```
Right, that's what we're doing in this procedure. I'm not clear on whether you want to make the intro text more explicit, or if you're commenting on the procedure below.

arthurevans replied:
Made text more explicit.

arthurevans commented on packages/lit-dev-content/site/blog/2022-07-19-lit-for-polymer-users.md:
```diff
@@ -0,0 +1,935 @@
+
+#### Replacing two-way bindings
+
+There are many ways to replace two-way bindings. On the more complex side, you can use a state container like Redux. Individual components can subscribe for updates to the state, and dispatch actions to update the state. This is a great way to go, but it may require a lot of work.
```
Hmm... If I just put the redux comment at the end it's going to be totally lost, and it's not enough text to be its own section. Maybe I can add it as an aside near the end.

arthurevans replied:
Done.

arthurevans commented on packages/lit-dev-content/site/blog/2022-07-19-lit-for-polymer-users.md:
```diff
@@ -0,0 +1,935 @@
+
+This last item is the most problematic. Components fire change events for any change to a property, and each change event is handled synchronously.  Widespread use of two-way binding across an entire application can make it hard to reason about data flow and the order in which components update themselves. A change can propagate both up and down the tree, potentially leading to components being u ...
+
+In general the Lit team recommends using top-down data flow using immutable data patterns, which make it easier to reason about how a data change propagates. To avoid issues, components should *not* generate an event when a property is set from above (that is, set programmatically from outside the component). A component should generate events:
```
Done. Reordered this as an intro to the bulleted list, to make it clear that we're basically summarizing the best practices.

arthurevans commented on packages/lit-dev-content/site/blog/2022-07-19-lit-for-polymer-users.md:
````diff
@@ -0,0 +1,935 @@
+<target-element @click=${handler}></target-element>
+```
+
+In Lit, the handler can be either an unbound method, like `${this.clickHandler}` or an arrow function. Using an arrow function, you can close over other data or call a function with a different signature. For example:
````
Removed "unbound" (also changed "an" => "a" so doing the change manually, not committing this suggestion).

arthurevans commented on packages/lit-dev-content/site/blog/2022-07-19-lit-for-polymer-users.md:
````diff
@@ -0,0 +1,935 @@
+
+Polymer:
+
+```js
````
Fixed missing code fence.

arthurevans commented on packages/lit-dev-content/site/blog/2022-07-19-lit-for-polymer-users.md:
````diff
@@ -0,0 +1,935 @@
+}
+```
+
+In Lit, you provide styles in a static `styles` field using the `css` tag function.
````
Noted.

arthurevans commented on packages/lit-dev-content/site/blog/2022-07-19-lit-for-polymer-users.md:
```diff
@@ -0,0 +1,935 @@
+
+Like Polymer, Lit supports setting properties, attributes, and event handlers using expressions. Lit uses slightly different syntax, with prefixes instead of suffixes.
+
+Polymer:
```
Done. Ideally we'd make the table horizontally scrollable on small screens, because it's a little awkward there, but it does work.

arthurevans commented on packages/lit-dev-content/site/blog/2022-07-19-lit-for-polymer-users.md:
```diff
@@ -0,0 +1,935 @@
+
+### Data binding
+
+Where Polymer has data binding, Lit has _expressions_ in its templates. Lit expressions are standard JavaScript expressions that are bound to a particular location in the template. As such, Lit expressions can do almost everything you can do with Polymer data bindings, and many things that you can't easily do in Polymer.
```
Going to leave this one for now. Felt like the alternative was a little bit less clear than the original.

arthurevans commented on packages/lit-dev-content/site/blog/2022-07-19-lit-for-polymer-users.md:
````diff
@@ -0,0 +1,935 @@
+In Lit, you can use a JavaScript conditional expressions. The conditional operator (or ternary) works well:
+
+```js
+html`${this.condition
````
I thought we'd decided we should use `map` by default from now on, because it has some advantages over array.map (namely, being able to use other, non-array iterables).

Whereas `when` doesn't seem to add anything other than another import statement.

arthurevans commented on packages/lit-dev-content/site/blog/2022-07-19-lit-for-polymer-users.md:
```diff
@@ -0,0 +1,935 @@
+
+By default, Polymer's `dom-if` behaves a little differently from a Lit conditional. When the condition goes from a truthy to a falsy value, the `dom-if` simply _hides_ the conditional DOM, instead of removing it from the DOM tree. This may save some resources when the condition becomes truthy again.
+
+On the other hand, Lit removes and discards the DOM when a condition changes; `dom-if` does the same thing if you set the `restamp` property to `true`.
```
Added (and rewrote section a good bit to try and explain the differences between `hidden=${condition}`, a simple JavaScript conditional, and the `cache` directive.)

arthurevans commented:
Addressed feedback, I think.

Once this doc looks GTG, I'll close this PR and open a new one based on top of Elliott's new article PR. I've tested this out, and it looks pretty good.

arthurevans commented:
Ping @sorvell

arthurevans commented:
Now replaced by #909. 

Leaving this as a draft for now in case anyone wants to check the comments/diffs.

sorvell commented on packages/lit-dev-content/site/blog/2022-07-19-lit-for-polymer-users.md:
```diff
@@ -0,0 +1,1013 @@
+
+This last item is the most problematic. Components fire change events for any change to a property, and each change event is handled synchronously.  Widespread use of two-way binding across an entire application can make it hard to reason about data flow and the order in which components update themselves.
+
+Ideally an event is a discrete signal sent to communicate an explicit change that isn't otherwise easily observable. Sending an event as a side-effect of setting a property as Polymer makes the communication potentially redundant and implicit. This implicit behavior, in particular, can make data flow hard to understand.
```
```suggestion
Ideally an event is a discrete signal sent to communicate an explicit change that isn't otherwise easily observable. Sending an event as a side-effect of setting a property as Polymer does makes the communication potentially redundant and implicit. This implicit behavior, in particular, can make data flow hard to understand.
```

sorvell commented on packages/lit-dev-content/site/blog/2022-07-19-lit-for-polymer-users.md:
```diff
@@ -0,0 +1,1013 @@
+
+Ideally an event is a discrete signal sent to communicate an explicit change that isn't otherwise easily observable. Sending an event as a side-effect of setting a property as Polymer makes the communication potentially redundant and implicit. This implicit behavior, in particular, can make data flow hard to understand.
+
+To summarize the [Custom Element Best Practices](https://web.dev/custom-elements-best-practices/#events) guidelines, a component should* fire events:
```
```suggestion
To summarize the [Custom Element Best Practices](https://web.dev/custom-elements-best-practices/#events) guidelines, a component *should* fire events:
```

sorvell commented on packages/lit-dev-content/site/blog/2022-07-19-lit-for-polymer-users.md:
```diff
@@ -0,0 +1,1013 @@
+To summarize the [Custom Element Best Practices](https://web.dev/custom-elements-best-practices/#events) guidelines, a component should* fire events:
+
+*   When the state of the element changes as a result of user interaction—like clicking a button or editing a text field inside the component.
+*   When something internal changes inside the component—like a timer going off or an animation completing.
```
```suggestion

*   When something internal changes inside the component—like a timer going off or an animation completing.
```

sorvell reviewed (approved):
Just a couple of very small typo fixes but otherwise looks good!

arthurevans commented:
Cool! I'll make those changes on #909.

arthurevans commented:
Fixed in https://github.com/lit/lit.dev/pull/909/commits/c3398eae897a7f0256e362fbda309474de6aa299
